File: src/env_setup.py
"""Provision the runtime environment for a project."""

import os
import subprocess
import shutil

import anthropic
from rich.console import Console
from rich.progress import Progress

from src.analyzer import ProjectInfo

console = Console()


class EnvSetup:
    def __init__(self, api_key: str, work_dir: str = ".devbox"):
        self.client = anthropic.Anthropic(api_key=api_key)
        self.work_dir = work_dir

    def provision(self, project: ProjectInfo) -> None:
        repo_path = os.path.join(self.work_dir, project.name)

        console.print(f"[bold] Setting up environment for {project.name}[/bold]")
        console.print(f"  Stack: {project.stack}")
        console.print(f"  Runtime: {project.runtime}")

        # Check what's available locally
        available = self._check_available_tools()

        # Use AI to generate the best setup plan given local tools
        setup_plan = self._generate_setup_plan(project, available)

        with Progress() as progress:
            task = progress.add_task("[cyan]Provisioning...", total=len(setup_plan))

            for step in setup_plan:
                console.print(f"  -> {step['description']}")
                try:
                    subprocess.run(
                        step["command"],
                        shell=True,
                        check=True,
                        cwd=repo_path,
                        capture_output=True,
                    )
                except subprocess.CalledProcessError as e:
                    console.print(f"[red] Failed: {e.stderr.decode()}[/red]")
                    # Ask AI how to recover
                    fix = self._suggest_fix(project, step, e)
                    console.print(f"[yellow]  Suggestion: {fix}[/yellow]")

                progress.update(task, advance=1)

        # Write .env if needed
        if project.env_vars:
            self._write_env_template(repo_path, project.env_vars)

        console.print("[green] Environment ready.[/green]")

    def _check_available_tools(self) -> dict[str, bool]:
        tools = {
            "docker": shutil.which("docker"),
            "node": shutil.which("node"),
            "npm": shutil.which("npm"),
            "python3": shutil.which("python3"),
            "pip": shutil.which("pip"),
            "go": shutil.which("go"),
            "rust": shutil.which("rustc"),
            "ruby": shutil.which("ruby"),
            "java": shutil.which("java"),
            "maven": shutil.which("mvn"),
            "gradle": shutil.which("gradle"),
        }
        return {k: bool(v) for k, v in tools.items()}

    def _generate_setup_plan(self, project: ProjectInfo, available: dict) -> list[dict]:
        prompt = f"""Generate a setup plan for this project. Available tools: {available}.
Project stack: {project.stack}
Runtime: {project.runtime}
Dependencies: {project.dependencies}

Return a JSON list of steps, each with "description" and "command" keys.
Use Docker if native tools are missing. Prefer native tools when available.
Only output the JSON array, nothing else."""

        response = self.client.messages.create(
            model="claude-sonnet-4-6",
            max_tokens=1024,
            messages=[{"role": "user", "content": prompt}],
        )

        import json
        text = response.content[0].text
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            return [{"description": f"Install dependencies", "command": "echo 'No steps generated'"}]

    def _suggest_fix(self, project: ProjectInfo, step: dict, error: Exception) -> str:
        response = self.client.messages.create(
            model="claude-sonnet-4-6",
            max_tokens=256,
            messages=[{
                "role": "user",
                "content": f"Setup step failed: {step}\nError: {error}\nProject: {project.stack}\nSuggest a fix in one sentence."
            }],
        )
        return response.content[0].text

    def _write_env_template(self, repo_path: str, env_vars: list[str]) -> None:
        env_path = os.path.join(repo_path, ".env")
        if os.path.exists(env_path):
            return
        with open(env_path, "w") as f:
            for var in env_vars:
                f.write(f"{var}=changeme\n")
        console.print(f"  Created .env with {len(env_vars)} variables to configure")

File: src/main/java/com/devbox/env/EnvSetup.java
package com.devbox.env;

import com.devbox.model.ProjectInfo;
import com.google.gson.*;
import com.google.gson.reflect.TypeToken;

import java.io.*;
import java.lang.reflect.Type;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.TimeUnit;

public class EnvSetup {
    private static final String API_URL = "https://api.anthropic.com/v1/messages";
    private static final String MODEL = "claude-sonnet-4-6-20250514";

    private final String apiKey;
    private final String workDir;
    private final HttpClient httpClient;
    private final Gson gson;

    public EnvSetup(String apiKey, String workDir) {
        this.apiKey = apiKey;
        this.workDir = workDir;
        this.httpClient = HttpClient.newHttpClient();
        this.gson = new Gson();
    }

    public void provision(ProjectInfo project) throws Exception {
        Path repoPath = Paths.get(workDir, project.getName());

        System.out.println("⚙ Setting up environment for " + project.getName());
        System.out.println("  Stack: " + project.getStack());
        System.out.println("  Runtime: " + project.getRuntime());

        Map<String, Boolean> available = checkAvailableTools();
        List<Step> plan = generateSetupPlan(project, available);

        for (Step step : plan) {
            System.out.println("  -> " + step.description);
            try {
                ProcessBuilder pb = new ProcessBuilder("sh", "-c", step.command);
                pb.directory(repoPath.toFile());
                pb.redirectErrorStream(true);
                Process p = pb.start();
                boolean ok = p.waitFor(120, TimeUnit.SECONDS);
                if (!ok || p.exitValue() != 0) {
                    String output = new String(p.getInputStream().readAllBytes());
                    System.out.println("  ⚠ Command may have issues: " + output.substring(0, Math.min(200, output.length())));
                }
            } catch (Exception e) {
                String suggestion = suggestFix(project, step, e.getMessage());
                System.out.println("  💡 Suggestion: " + suggestion);
            }
        }

        if (!project.getEnvVars().isEmpty()) {
            writeEnvTemplate(repoPath, project.getEnvVars());
        }

        System.out.println("✓ Environment ready.");
    }

    private Map<String, Boolean> checkAvailableTools() {
        String[] tools = {"docker", "node", "npm", "python3", "pip", "pip3",
                "go", "rustc", "ruby", "java", "mvn", "gradle"};
        Map<String, Boolean> result = new LinkedHashMap<>();
        for (String tool : tools) {
            try {
                Process p = new ProcessBuilder("which", tool).start();
                result.put(tool, p.waitFor() == 0);
            } catch (Exception e) {
                result.put(tool, false);
            }
        }
        return result;
    }

    private List<Step> generateSetupPlan(ProjectInfo project, Map<String, Boolean> available) throws Exception {
        String prompt = String.format("""
                Generate a setup plan for this project. Available tools on this machine: %s.
                Project stack: %s
                Runtime: %s
                Dependencies: %s

                Return ONLY a JSON array. Each element has "description" and "command" keys.
                Use Docker if native tools are missing. Prefer native tools when available.
                Use pip3 instead of pip if needed. Use npm/npx for Node projects.
                """,
                available, project.getStack(), project.getRuntime(), project.getDependencies()
        );

        String response = callAnthropic(prompt);
        // Parse JSON array
        String json = response.trim();
        if (json.startsWith("```")) {
            json = json.replaceAll("```json?|```", "").trim();
        }

        try {
            Type listType = new TypeToken<List<Step>>(){}.getType();
            return gson.fromJson(json, listType);
        } catch (Exception e) {
            return List.of(new Step("Install dependencies", "echo 'No setup steps generated'"));
        }
    }

    private String suggestFix(ProjectInfo project, Step step, String error) throws Exception {
        String prompt = String.format(
                "Setup step failed: %s\nError: %s\nProject: %s\nSuggest a fix in one sentence.",
                step.description, error, project.getStack()
        );
        return callAnthropicShort(prompt);
    }

    private void writeEnvTemplate(Path repoPath, List<String> envVars) throws IOException {
        Path envPath = repoPath.resolve(".env");
        if (Files.exists(envPath)) return;
        StringBuilder sb = new StringBuilder();
        for (String v : envVars) {
            sb.append(v).append("=changeme\n");
        }
        Files.writeString(envPath, sb.toString());
        System.out.println("  Created .env with " + envVars.size() + " variables to configure");
    }

    private String callAnthropic(String prompt) throws Exception {
        JsonObject body = new JsonObject();
        body.addProperty("model", MODEL);
        body.addProperty("max_tokens", 1024);
        body.add("messages", gson.toJsonTree(List.of(
                Map.of("role", "user", "content", prompt)
        )));

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(API_URL))
                .header("x-api-key", apiKey)
                .header("anthropic-version", "2023-06-01")
                .header("content-type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(gson.toJson(body)))
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        JsonObject json = JsonParser.parseString(response.body()).getAsJsonObject();
        return json.getAsJsonArray("content").get(0).getAsJsonObject().get("text").getAsString();
    }

    private String callAnthropicShort(String prompt) throws Exception {
        JsonObject body = new JsonObject();
        body.addProperty("model", MODEL);
        body.addProperty("max_tokens", 256);
        body.add("messages", gson.toJsonTree(List.of(
                Map.of("role", "user", "content", prompt)
        )));

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(API_URL))
                .header("x-api-key", apiKey)
                .header("anthropic-version", "2023-06-01")
                .header("content-type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(gson.toJson(body)))
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        JsonObject json = JsonParser.parseString(response.body()).getAsJsonObject();
        return json.getAsJsonArray("content").get(0).getAsJsonObject().get("text").getAsString();
    }

    static class Step {
        String description;
        String command;

        Step(String description, String command) {
            this.description = description;
            this.command = command;
        }
    }
}
